core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala

```diff
@@ -198,24 +198,70 @@ class QualificationAppInfo(
     }
   }
 
-  private def getStageToExec(execInfos: Seq[ExecInfo]): (Map[Int, Seq[ExecInfo]], Seq[ExecInfo]) = {
-    val execsWithoutStages = new ArrayBuffer[ExecInfo]()
-    val perStageSum = execInfos.flatMap { execInfo =>
+  private def checkStageIdInExec(prev: Option[ExecInfo],
+      execInfo: ExecInfo, next: Option[ExecInfo]): (Seq[(Int, ExecInfo)], Option[ExecInfo]) = {
+    val associatedStages = {
       if (execInfo.stages.size > 1) {
-        execInfo.stages.map((_, execInfo))
+        execInfo.stages.toSeq
       } else if (execInfo.stages.size < 1) {
-        // we don't know what stage its in or its duration
-        logDebug(s"No stage associated with ${execInfo.exec} " +
-          s"so speedup factor isn't applied anywhere.")
-        execsWithoutStages += execInfo
-        Seq.empty
+        if (prev.exists(_.stages.size >= 1)) {
+          prev.flatMap(_.stages.headOption).toSeq
+        } else if (next.exists(_.stages.size >= 1)) {
+          next.flatMap(_.stages.headOption).toSeq
+        } else {
+          // we don't know what stage its in or its duration
+          logDebug(s"No stage associated with ${execInfo.exec} " +
+            s"so speedup factor isn't applied anywhere.")
+          Seq.empty
+        }
+      } else {
+        Seq(execInfo.stages.head)
+      }
+    }
+    if (associatedStages.nonEmpty) {
+      (associatedStages.map((_, execInfo)), None)
+    } else {
+      (Seq.empty, Some(execInfo))
```

Comment:
This block does the processing when `execInfo.stages` is not empty, else it returns a seq of `execInfo.stages`.

Can we merge the blocks for `if (execInfo.stages.size > 1)` and `else` case?